mllib/src/main/scala/org/apache/spark/mllib/clustering/DpMeans.scala

```diff
@@ -0,0 +1,247 @@
+
+        val sums = Array.fill(k)(Vectors.zeros(dims))
+        val counts = Array.fill(k)(0L)
+        val totalCost = Array.fill(k)(0.0D)
+
+        points.foreach { point =>
+          val (currentCenter, cost) = DpMeans.assignCluster(activeCenters, point)
+          totalCost(currentCenter) += cost
+          val currentSum = sums(currentCenter)
+          axpy(1.0, point.vector, currentSum)
+          counts(currentCenter) +=1
+        }
+
+        val result = for (i <- 0 until k) yield {
+          (i, (sums(i), counts(i)))
+        }
+        result.iterator
+      }.reduceByKey(mergeClusters).collectAsMap()
+
+      // Update the cluster centers
+      var changed = false
+      var j = 0
+      val currentK = clusterStat.size
+      while (j < currentK) {
+        val (sumOfPoints, count) = clusterStat(j)
+        if (count != 0) {
+          scal(1.0 / count, sumOfPoints)
+          val newCenter = new VectorWithNorm(sumOfPoints)
+          // Check for convergence
+          globalCenters.length match {
+            case currentK => if (DpMeans.squaredDistance(newCenter, globalCenters(j)) >
+              convergenceTol * convergenceTol) {
+              changed = true
+              }
+            case _ => changed = true
+          }
+          globalCenters(j) = newCenter
+        }
+        j += 1
+      }
+      if (!changed) {
+        converged = true
+        logInfo("DpMeans clustering finished in " + (iteration + 1) + " iterations")
+      }
+      iteration += 1
+      norms.unpersist()
+    }
+
+    if (iteration == maxIterations) {
+      logInfo(s"DPMeans reached the max number of iterations: $maxIterations.")
+    } else {
+      logInfo(s"DPMeans converged in $iteration iterations")
+    }
+    new DpMeansModel(globalCenters.toArray.map(_.vector))
+  }
+}
+/**
+ * Core methods of  DP means clustering.
+ */
+private object DpMeans {
+  /**
+   * A data point is said to be "covered" by a cluster `c` if the distance from the point
+   * to the cluster center of `c` is less than a given lambda value.
+   */
+  def cover(
+      points: Iterator[VectorWithNorm],
+      centers: ArrayBuffer[VectorWithNorm],
+      lambda: Double): Iterator[VectorWithNorm] = {
+    var newCenters = ArrayBuffer.empty[VectorWithNorm]
+    if(!points.isEmpty){
+      if (centers.length == 0) newCenters += points.next
+      points.foreach { z =>
+        val dist = newCenters.union(centers).map { center => squaredDistance(z, center) }
+        if (dist.min > lambda) newCenters += z
+      }
+    }
+    newCenters.iterator
+  }
+
+  /**
+   * Each data point is assigned to the nearest cluster. This method returns
+   * the corresponding cluster label and the distance from the center to the
+   * cluster center.
+   */
+    def assignCluster(
+        centers: ArrayBuffer[VectorWithNorm],
+        x: VectorWithNorm): (Int, Double) = {
+      val dist = centers.map { c => squaredDistance(x, c) }
+      (dist.indexOf(dist.min), dist.min)
```

Comment:
@mengxr Do you want me to reuse the findClosest function in Kmeans instead of defining something similar here?